Use extras_require in setup.py for optional dependencies

mivade opened the pull request:
Include `requests` and `aiohttp` as both appear to be required for HTTP support now.

This relates to comments in #366 

Are there any other optional dependencies we should add here? Once those are added I would also update the documentation to include hints about this, so I'm marking this as a draft PR for now.

martindurant commented:
Yes, quite a few! Any of the items in registry that are imported from within the fsspec package and have error messages, e.g., distributed for dask, paramiko for ssh, pyarrow for hdfs.

mivade commented:
Thanks @martindurant, I'll take a look there.

martindurant commented on setup.py:
```diff
@@ -32,5 +32,19 @@
     python_requires=">3.6",
     install_requires=open("requirements.txt").read().strip().split("\n"),
+    extras_require={
+        "aws": ["s3fs"],
```
I would probably use the same names as the protocols, so "s3" for this one and "gcs" instead of "google"

mivade replied:
I wasn't sure what the best approach was there. I think matching it is fine even if that means there are some redundancies.

martindurant commented on docs/source/index.rst:
```diff
@@ -31,6 +31,11 @@ Installation
+Not all included filesystems are usable by default without installing extra
+dependencies. For example to be able to access data in S3::
+
+   pip install fsspec[s3]
```
Missing s3 in the list :)

martindurant replied:
Please also check that this works - because s3fs depends on fsspec, fsspec[s3] would depend on itself.

martindurant commented:
This PR is still marked as draft. Ready to go?
I have not personally checked that something like `pip install fsspec[s3]` would do the right thing.

mivade commented:
Not yet. I still want to add automated testing that the extras work. I'll mark it as ready for review once I've done so.

martindurant commented:
ping @mivade , do you think you'll have time to finish this PR off?

mivade commented:
Yeah, I should be able to get to the testing later this week. Sorry for the delay!

mivade commented:
@martindurant I've added some pretty basic testing here. I took the existing S3 and GCSFS tox environments and removed the conda dependencies installed there to instead use the `extras=` line. This tells tox to install the package with the provided extras (e.g., something like `pip install .[gcs]`). Of course this doesn't test every single extra defined in this PR; doing so would require some considerable rework of how the test environments are setup.

martindurant commented:
Looks good!
